micromanager_gui/main_window.py

```diff
@@ -152,19 +175,126 @@ def add_frame_explorer(self, name, array):
+                break
+            else:
+                val += 1
+        return fname
+
+    def _on_mda_finished(self, sequence: useq.MDASequence):
+        """Save layer and add increment to save name."""
+        
+        if self.mda.save_groupBox.isChecked():
+
+            try:
+                active_layer = next(l for l in self.viewer.layers if l.metadata.get('uid') == sequence.uid)
+            except StopIteration:
+                raise IndexError("could not find layer corresponding to sequence")
+
+            fname =  self.mda.fname_lineEdit.text()
+
+            save_path = Path(self.mda.dir_lineEdit.text())
+
+            """Look into the saving directory (save_path) and get tif files names"""
+
+            list_dir = [f.name for f in save_path.glob("*.tif")]
+
+            """
+               Ckeck the input filename and modify it to add _nnn in the end.
+               The get_filename(fname,list_dir) function check if the name 
+               is already present in the save_path and incremant the _nnn accordingly.
+
+               filename examples: user input -> output:
+                - mda       -> mda_000
+                - mda_3     -> mda_003
+                - mda_0001  -> mda_001
+                - mda1      -> mda_001
+                - mda011021 -> mda011021_000
+                - mda_011021 -> mda_011021_000
+            """
+            try:
+                n = fname.split('_')[-1]
+                int_n = int(n)
+                
+                if len(n) == 3 and int_n >= 0:
+                    fname = self.get_filename(fname,list_dir)
+                
+                elif len(n) != 3 and len(n) <=4 and int_n >= 0:
+                    s = ''
+                    for i in fname.split('_')[:-1]:
+                        s = s + i + '_'
+                        print(s)
+                    print('s', s)
+                    fname = s + '{0:03}'.format(int_n)
+                    fname = self.get_filename(fname,list_dir)
+                
+                else:
+                    fname = fname + '_000'
+                    fname = self.get_filename(fname,list_dir)
+
+            except ValueError:
+                n = ''
+                for i in range(1,len(fname)+1):
+                    try:
+                        n += str(int(fname[-i]))
+                    except ValueError:
+                        break
+                if len(n) > 0 and len(n) <= 4:
+                    n = n[::-1]
+                    fname = fname.replace(n, '_' + '{0:03}'.format(int(n)))
+                    fname = self.get_filename(fname,list_dir)
+                else:
+                    fname = fname + '_000'
+                    fname = self.get_filename(fname,list_dir)
+
+
+            #TODO: save also metadata
+            self.viewer.layers[str(active_layer)].save(str(save_path / fname))
+
+            """update filename in mda.fname_lineEdit for the next aquisition."""
```

Comment:
use `#` type comment

Reply:
also ... the `list_dir` variable will go out of scope at the end of this function (i.e. it's not `self.list_dir`) ... so you're not actually updating anything for the next acquisition here.  In any case though ... getting a list of files in a directory is a fast operation, so it's not worth maintaining a separate internal list that has the potential to go out of sync with the folder.  just check it every time